Add InsertAllRequest, InsertAllResponse and model classes for table data

mziccard opened the pull request:
This PR will probably require some discussion :) I am eager to receive any feedback.
It adds:
- `InsertAllRequest` and `InsertAllResponse` to insert data using [TableData: insertAll](https://cloud.google.com/bigquery/docs/reference/v2/tabledata/insertAll).
- `TableRow` and `FieldValue` to hold the results of [TableData: list](https://cloud.google.com/bigquery/docs/reference/v2/tabledata/list) and [query](https://cloud.google.com/bigquery/docs/reference/v2/jobs/query)

Remarks:
- We discussed with @aozarov about adding more information to [FieldValue.Kind](https://github.com/mziccard/gcloud-java/blob/20dfb487637d8d31903a352e443867e505c3ac06/gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java#L30) as `REPEATED_OF_PRIMITIVE` and `REPEATED_OF_RECORD` but I realized that an empty repeated field is returned as an empty array and there's no way to say whether it is of primitive or record type.

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,207 @@
+
+/**
+ * Google BigQuery Table Field Value class. Objects of this class represent values of a BigQuery
+ * Table Field. A list of values forms a {@link TableRow}. Tables rows can be get as the result of
```
s/can be get/can be gotten/g

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/TableRowTest.java:
```diff
@@ -0,0 +1,46 @@
+package com.google.gcloud.bigquery;
```
Nit: license statement missing

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,207 @@
+package com.google.gcloud.bigquery;
```
Nit: license statement missing

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,119 @@
+
+  /**
+   * Returns all insertion errors as a map whose keys are indexes of rows that failed to insert. To
+   * each failed row index a non-empty list of {@link BigQueryError} is associated.
```
Nit: I think this sentence would sound slightly better as: "Each failed row index is associated with a non-empty list of {@link BigQueryError}s"

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,119 @@
+ * {@link InsertAllRequest}. If a row failed to be inserted, the non-empty list of errors associated
+ * to that row's index can be obtained with {@link InsertAllResponse#errorsFor(Long)}.
+ * {@link InsertAllResponse#insertErrors()} can be used to return all errors caused by a
+ * {@link InsertAllRequest}, as a map.
```
nit: can remove comma

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  /**
+   * Returns {@code true} if no row insertion failed, false otherwise. If {@code true}
```
At the end of this line, should "true" be "false"?

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,207 @@
+
+  /**
+   * Return this field's value as an {@link String}.
+   */
```
Should this javadoc have: "This method should only be used if the corresponding field has {@link Field.Type#string}" warning and `@throws` statement?

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllRequest.java:
```diff
@@ -0,0 +1,400 @@
+     * @param id id of the row, used to identify duplicates
+     * @param content the actual content of the row
+     */
+    public RowToInsert(String id, Map<String, Object> content) {
```
Based off previous code, it looks like `gcloud-java` has the precedent of using `of` in this type of situation rather than expose the constructor.

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/FieldValueTest.java:
```diff
@@ -0,0 +1,101 @@
+      ImmutableMap.<String, Object>of("f", ImmutableList.<Object>of(FLOAT_FIELD, TIMESTAMP_FIELD));
+
+  @Test
+  public void testFromPb() {
```
Should there be a check for a FieldValue with value `null`?

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/FieldValueTest.java:
```diff
@@ -0,0 +1,101 @@
+public class FieldValueTest {
+
+  private static final TableCell BOOLEAN_FIELD = new TableCell().setV("false");
+  private static final Map<String, String> INTEGER_FIELD = ImmutableMap.of("v", "1");
```
Any reason these aren't `new TableCell().setV(...)`?

mziccard replied:
I am trying to test different types that can be passed to `fromPb`. First level field values (in a row) are passed as a `TableCell` object, while field values nested in a record are just an `ArrayMap` hence I am using `instance of Map` in `fromPb`.

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableRow.java:
```diff
@@ -0,0 +1,104 @@
+    return obj instanceof TableRow && Objects.equals(values, ((TableRow) obj).values);
+  }
+
+  static TableRow fromPb(com.google.api.services.bigquery.model.TableRow rowPb) {
```
I think we should add a test for this method in TableRowTest

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/TableRowTest.java:
```diff
@@ -0,0 +1,46 @@
+  private static final FieldValue TIME_VALUE = new FieldValue(FieldValue.Kind.PRIMITIVE, "42");
+  private static final List<FieldValue> ROW_VALUES =
+      ImmutableList.of(BOOL_VALUE, INT_VALUE, FLOAT_VALUE, STRING_VALUE, TIME_VALUE);
+  private static final TableRow TABLE_ROW1 = new TableRow.Builder()
```
Could there theoretically be `FieldValue`s with value `null` in a TableRow?  If so I think we should add a `NULL_VALUE`

ajkannan replied:
Looking a bit more, it seems like TableRows can in fact contain null FieldValues.  Am I wrong in my understanding (or was not adding a `NULL_VALUE` to this table row an accidental leave-out)?  Also it might be nice to add a `repeatedValue` and a `recordValue` as well.

mziccard replied:
You are right, it can contain null values as well as repeated and record, but what's the point of adding them too? I mean: we are unit testing `TableRow` which is merely a collection of field values and we already unit-tested `FieldValue` so I don't think this is really needed. But I'll add them anyway.

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/InsertAllRequestTest.java:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  @Test
+  public void testOf() {
```
nit: I think two of the `of` methods with varargs in the method signature aren't tested

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/InsertAllRequestTest.java:
```diff
@@ -0,0 +1,161 @@
+      ImmutableList.of(new InsertAllRequest.RowToInsert("id1", CONTENT1),
+          new InsertAllRequest.RowToInsert("id2", CONTENT2));
+  private static final TableId TABLE_ID = TableId.of("dataset", "table");
+  private static final boolean SKIP_INVALID_ROWS = true;
```
Nit: for the Boolean values, I think it would be nice to vary the values.  Say someone in the future accidentally always sets SKIP_INVALID_ROWS to true or IGNORE_UNKNOWN_VALUES to false, then this test wouldn't catch that.

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/InsertAllRequestTest.java:
```diff
@@ -0,0 +1,161 @@
+  private static final TableId TABLE_ID = TableId.of("dataset", "table");
+  private static final boolean SKIP_INVALID_ROWS = true;
+  private static final boolean IGNORE_UNKNOWN_VALUES = false;
+  private static final InsertAllRequest INSERT_ALL_REQUEST1 = InsertAllRequest.builder(TABLE_ID)
```
Nit: one of the 8 builders used here could use the `BaseTableInfo` parameter

mziccard commented:
Thanks @ajkannan for the pass. I fixed the comments in place.

mziccard commented:
Should we make all value getters in `FieldValue` throw the same exception? Right now they either throw `ClassCastException`, `IllegalStateException`, `NumberFormatException` or `NullPointerException` depending on the type of conversion being asked. What do you think?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,232 @@
+  /**
+   * The field value's kind, giving information on the field's content type.
+   */
+  public enum Kind {
```
I wish we could find some other name (as kind is used for a different purpose in the underlying layer ). I am not sure what though. `Type`, `TypeAttributes`, ..

mziccard replied:
I did not pick `Type` to avoid confusion with `Field.Type` but you are right that kind is also overloaded. I couldn't come up with a better idea though

aozarov replied:
`Attribute`?

mziccard replied:
I'm on the fence on this so if `Attribute` is your preference I am more than ok to go with it.

ajkannan replied:
I think I would slightly prefer `Attribute`.  Just throwing out another option: we could change `Field.Type` to `Field.PrimitiveType`/ `Field.DataType` and then changed `Kind` to `FieldType`.  Having Field.FieldType is a little repetitive, but I couldn't really think of anything better.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,232 @@
+   * Returns this field's value as a {@link Long}. This method should only be used if the
+   * corresponding field has {@link Field.Type#integer()} type.
+   *
+   * @throws NumberFormatException if the field's value could not be converted to {@link Integer}
```
here and below - could also throw ClassCastException

mziccard replied:
Fixed in 1674c9c

aozarov replied:
Do you think we should add a checkNotNull in `stringValue()` (and remove in other places such as `booleanValue`) to make it consistent (get value methods throw NPE if `isNull() == true`)?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,236 @@
+   * {@link Field.Type#floatingPoint()}, {@link Field.Type#integer()},
+   * {@link Field.Type#timestamp()}).
+   *
+   * @throws ClassCastException if the field has not primitive type
```
here and below - is it "has not" or "is not a"?

mziccard replied:
Mmmm do you think there's a big difference here? I don't mind changing

ajkannan replied:
I think "is not a" sounds better, I support changing it

aozarov replied:
Nit, but I think the same applies to the description part.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,236 @@
+  @SuppressWarnings("unchecked")
+  public boolean booleanValue() {
+    String stringValue = stringValue();
+    checkState(stringValue.equalsIgnoreCase("true") || stringValue.equalsIgnoreCase("false"),
```
I assume because we are using the json format it is always going to be "true" or "false" and this class will never be used for a CSV format which also supports 1 or 0?

mziccard replied:
Yes, in JSON format booleans should only be "true" or "false"

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,236 @@
+   * Return this field's value as an {@link Object}.
+   */
+  public Object value() {
+    return value;
```
what about handling nulls?

mziccard replied:
What do you mean with *handling nulls*?

aozarov replied:
I assume `value` could be `null` for primitive value or record.

If so, should we have `isNull` method and should we indicate that NPE will be thrown
from the xxxValue methods if `isNull` is `true`?

mziccard replied:
Oh sure I can add that.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,236 @@
+   */
+  @SuppressWarnings("unchecked")
+  public long timestampValue() {
+    return new Double(((Double.valueOf(stringValue())) * MICROSECONDS)).longValue();
```
So timestamp are always returned in json as a long (time since epoch in seconds)?

mziccard replied:
They are returned as a double value since epoch in the form `1408452095.22` where the integer part is **seconds**: `1408452095.22 == 2014-08-19 07:41:35.220 -05:00`

aozarov replied:
I see. Thanks for the clarification.

I think we should add this description as a comment in the code and also specify in the javadoc that this is "unix time" (or epoch time) .

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,236 @@
+   */
+  @SuppressWarnings("unchecked")
+  public long longValue() {
+    return Long.valueOf(stringValue());
```
Any reason you use `valueOf` (which returns `Long`) instead of `parseLong`?
Same for below.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/FieldValue.java:
```diff
@@ -0,0 +1,236 @@
+   */
+  public enum Kind {
+    /**
+     * A primitive field value. A {@code FieldValue} has type {@link #PRIMITIVE} when the
```
Nit,  `{@link #PRIMITIVE}` link to itself

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllRequest.java:
```diff
@@ -0,0 +1,399 @@
+  /**
+   * A Google Big Query row to be inserted into a table. Each {@code RowToInsert} has an associated
+   * id used by BigQuery to detect duplicate insertion requests on a best-effort basis.
+   *
```
Nit, here and below remove empty line before `<p>` tag.

mziccard replied:
That's the one time I do something to make checkstyle happy:) It was complaining with:
> \<p\> tag should be precede with empty line.

aozarov replied:
You are right! I am not sure why my IDE complaint about it. I guess it was configured with Sun style (which seems to avoid the [empty line](http://www.oracle.com/technetwork/articles/java/index-137868.html))?

Also, when looking at guava code it seems that `<p>` is indeed preceded by an empty line however never after (the paragraph starts on the same line).

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllRequest.java:
```diff
@@ -0,0 +1,399 @@
+     * <p>
+     * Example usage of adding a row with associated id:
+     * <pre>    {@code
+     *   InsertAllRequest.Buider builder = InsertAllRequest.builder(tableId);
```
s/Buider/Builder/g

mziccard replied:
Fixed in d2f0a82

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllRequest.java:
```diff
@@ -0,0 +1,399 @@
+
+    RowToInsert(String id, Map<String, Object> content) {
+      this.id = id;
+      this.content = content;
```
Should we copy the content?

mziccard replied:
Done in d2f0a82

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllRequest.java:
```diff
@@ -0,0 +1,399 @@
+     * @param content the actual content of the row
+     */
+    public static RowToInsert of(String id, Map<String, Object> content) {
+      return new RowToInsert(checkNotNull(id), checkNotNull(content));
```
I thought `id` can be `null` or did you plan to provide another version of `of(Map<String, Object> content)` without id?

mziccard replied:
My idea was that if `RowToInsert` is used an `id` *must* be provided. If the user wants to insert a row without id `addRow(Map<String, Object> content)` can be used.
However, I don't mind adding a new factory method if you prefer.

mziccard replied:
Added the missing factory method in d2f0a82

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,119 @@
+
+  TableDataInsertAllResponse toPb() {
+    TableDataInsertAllResponse responsePb = new TableDataInsertAllResponse();
+    if (insertErrors.size() > 0) {
```
Nit, `!insertError.isEmpty()`

mziccard replied:
Fixed in d2f0a82

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,119 @@
+ * {@link InsertAllResponse#insertErrors()} can be used to return all errors caused by a
+ * {@link InsertAllRequest} as a map.
+ */
+public class InsertAllResponse implements Serializable {
```
missing `serialVersionUID`

mziccard replied:
Fixed in d2f0a82

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,119 @@
+  /**
+   * Returns errors for the provided row index. If no error exists returns {@code null}.
+   */
+  public List<BigQueryError> errorsFor(Long index) {
```
why `Long`?

mziccard replied:
Fixed in d2f0a82

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,119 @@
+    return MoreObjects.toStringHelper(this).add("insertErrors", insertErrors).toString();
+  }
+
+  TableDataInsertAllResponse toPb() {
```
is that for testing only?

mziccard replied:
Yes, the conversions from and to `TableDataInsertAllResponse` are more complex than usual so I added this for easier testing. I could remove it though.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableRow.java:
```diff
@@ -0,0 +1,104 @@
+ * Google BigQuery Table Row class. Objects of this class contain a list of {@link FieldValue}, one
+ * for each field in the table.
+ */
+public class TableRow implements Serializable {
```
I am not totally against but I see no reason (or a potential benefit) for making users go an extra step (and call `values()`) for getting the values. Do you?

mziccard replied:
There are not practical benefit. As there were no particular ones in having the `Schema` class. I added this just to wrap each row into its own class.
If we remove this class we will have:
```java
Page<List<FieldValue>> results = bigquery.listTableData(tableId);
```
which is ok to me.

mziccard replied:
I removed the `TableRow` class in 50d381f

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllRequest.java:
```diff
@@ -0,0 +1,410 @@
+    /**
+     * Creates a row to be inserted without associated id.
+     *
+     * @param id id of the row, used to identify duplicates
```
extra `param`

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllResponse.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * Google Cloud BigQuery insert all response. Objects of this class possibly contain errors for an
+ * {@link InsertAllRequest}. If a row failed to be inserted, the non-empty list of errors associated
+ * to that row's index can be obtained with {@link InsertAllResponse#errorsFor(Long)}.
```
`{@link InsertAllResponse#errorsFor(long)}`

aozarov commented:
Looks great! Few nits.

mziccard commented:
Fixed all comments in 43eb31b. I added `checkNotNull` to `stringValue`, `repeatedValue` and `recordValue` and documented the NPE. I also documented that `value()` can return `null`. 43eb31b also fixes some other javadoc errors.
